### ISSUE TITLE
Script silently killed without error after assigning to script-scope var from imported function

### ISSUE THREAD

**Issue by lossycrypt** [bitburner-official/bitburner-src#642]
Version v2.3.2 aecdbe8, https://bitburner-official.github.io/bitburner-src/

I'm still new to BitBurner and Javascript, so I'm not sure if this is just some obscure quirk I don't understand yet.
I was trying to make a generic function ("waitExec") for waiting until a one-shot script has finished running, and noticed that the one-shot script was being killed for no apparent reason when importing waitExec from my library file, even though it worked fine when copying waitExec directly into the one-shot script. 

Demonstration: [bitburnerSave_1687436055_BN1x1.json.txt](https://github.com/bitburner-official/bitburner-src/files/11833797/bitburnerSave_1687436055_BN1x1.json.txt)

Steps:

1) On the termnial `run caller.js` will print
```
test.js: Test has been called!
caller.js: Testing local function: true
caller.js: Local test done.
```
2) Notice that it should also print this, but it _does not_.
```
test.js: Test has been called!
caller.js: Testing imported function: true
caller.js: Import test done.
```

3) Looking at Active Scripts -> Recently Killed one can see that both scripts were killed.

`tail: test.js`
```
Script finished running
exec: Failed to run due to script being killed.
```
`tail: caller.js`
```
exec: 'test.js' on 'home' with 1 threads and args: [].
sleep: Sleeping for 100 milliseconds
Script killed.
```

4) The killing seems to happen when test.js calls `initLib()` after caller.js has also called it. To my understanding the second call should be a quasi no-op. In my real game (v2.3.1) adding a reassignment check to initLib `export function initLib(NS){if(!ns){ns = NS}}` works, but not in this demo savegame.

5) After removing initLib() from test.js the call to testPrint() triggers one of those large concurrency error popups even though all ns.sleeps are being awaited?

Expected Behavior:
- I expected the scripts not to be killed. Heh.
- If it's normal that they're killed it would be nice if there was feedback i.e. an error message (popup or tail log).

Minimalistic reproduction setup:
`lib.js`
```
//pre-import ns so it doesn't have to be passed on every function call
var ns;
export function initLib(NS){ns = NS}

//waitExecLib is identical to waitExecCaller
export async function waitExecLib(script, host = 'home', threads = 1, ...args){
  const pid = ns.exec(script, host, threads, ...args)
  while (ns.isRunning(pid)){await ns.sleep(100)} 
  return (pid !== 0) // return false if the script never started (i.e. insufficient RAM)
  }

export function testPrint(msg){
  ns.tprint(msg)
}
```

`test.js`
```

import {initLib,testPrint} from 'lib.js'

/** @param {NS} NS */
export async function main(NS) {

  //Removing this line makes it not kill the scripts...
  initLib(NS);

  //...but then this causes a concurrency fail
  testPrint('Test has been called!')

}
```
`caller.js`
```

import {initLib,waitExecLib} from 'lib.js'

//waitExecLib is identical to waitExecCaller
async function waitExecCaller(script, host = 'home', threads = 1, ...args){
  const pid = ns.exec(script, host, threads, ...args)
  while (ns.isRunning(pid)){await ns.sleep(100)} 
  return (pid !== 0) // return false if the script never started (i.e. insufficient RAM)
  }

var ns;
/** @param {NS} NS */
export async function main(NS) {
  ns = NS;
  initLib(NS);

  //Always works
  ns.tprint('Testing local function: ', await waitExecCaller('test.js'))
  ns.tprint('Local test done.')

  //If test.js calls initLib(NS) this kills both caller.js and test.js
  ns.tprint('Testing imported function: ', await waitExecLib('test.js'))
  ns.tprint('Import test done.')

}
```

**Comment by lossycrypt** [comment]
On further through remembering that `ns` contains `ns.args` and is thus not identical for each call i can see that my approach might be flawed. But the silent killing is still strange.

**Comment by Snarling** [comment]
Yes, the `ns` sent in to each script is unique. It's how the game knows what script to compare for concurrency requirements (the same script can't do multiple things at once), how the game passes your function its arguments, how the game knows which script to assign ram cost for, etc.

Test.js is finishing normally, due to not having any awaited code and just running straight through. It's not being killed. That's why the exit message is `Script finished running` instead of `Script killed`.

Caller.js is being killed due to the second call to waitExecLib trying to use an `ns` instance that is no longer valid, which causes your script to error out.

> After removing initLib() from test.js the call to testPrint() triggers one of those large concurrency error popups even though all ns.sleeps are being awaited?

That's because you're attempting to call ns.tprint from the same ns instance that is already performing ns.sleep (from the other script). Use ns.asleep instead if you want to be able to do something else with the API during the sleep.

**Comment by lossycrypt** [comment]
Thank you very much for the detailed explanation. I'll try to wrap my head around it. I assumed that ns.exec would spawn a new procress, and thus assumed that caller.js and test.js would have seperate instances of lib.js. However what you're saying is that test.js - by it's own finishing - invalidates the ns instance passed to lib.js and caller.js is killed because it's using *the same* lib.js instance? So the `exec: Failed to run due to script being killed.` in test.js log *after* it finished is actually caused by caller.js calling `waitExecLib`?

Weirdly enough I've been using this NS passing construct all over the place and as long as I don't use ns.exec everything seems to work fine. Maybe I'm just lucky? Is there a method to verify if an ns instance is still valid?

> Caller.js is being killed due to the second call to waitExecLib trying to use an `ns` instance that is no longer valid, which causes your script to error out.

I would very much appreciate if that info - the reason for the killing - was somehow reflected in-game. Either in the log, or as a huge popup telling me to stop hitting myself ;)

**Comment by d0sboots** [comment]
Yeah, there's one other thing you were missing. Variables declared outside of any function are at *module level scope*. You can essentially think of them as global variables, although there are some slight differences. This is why `initLib` isn't working the way you want; I suspect you have it so you can avoid passing `ns` as an argument to every function, but instead it's getting overwritten.

What you actually want to be using for this case is classes, which *can* do what you want: take `ns` (and maybe other things) in the constructor and keep them in the class, where you can then refer to them through `this`.

**Comment by Snarling** [comment]
Not closing this just yet because it's possible we can improve this type of error so that more information is provided in the script log of the script that is dying, I just haven't had time to really look into it deeply yet.
